Bug: https://trac.sagemath.org/ticket/28856

From 447b6b071243f410df701679376c908394f759fd Mon Sep 17 00:00:00 2001
From: Timo Kaufmann <timokau@zoho.com>
Date: Mon, 9 Dec 2019 10:29:20 +0100
Subject: Make docbuild merging compatible with spinx 2.1

The attributes "todo_all_todos" and "citations" were previously
implicitly initialized by sphinx. That changed in sphinx 2.1 due to some
refactoring:

https://github.com/sphinx-doc/sphinx/commit/9abb4820b18201f63d77fdccb4ef394a21442f7a
https://github.com/sphinx-doc/sphinx/commit/885d35e374b56d1d17f5036fe07b74229cdbec7f

So now we need to check for the case where they are not initialized yet.
We can't actually update to sphinx 2.1 yet, since its python3 only. This
is mostly intended to make things easier for distros and allow us to
update in the future.
---
 src/sage_setup/docbuild/ext/multidocs.py | 19 ++++++++++++-------
 1 file changed, 12 insertions(+), 7 deletions(-)

From 9a98405a087758509af735b42939d6e91bf3dd68 Mon Sep 17 00:00:00 2001
From: Antonio Rojas <arojas@archlinux.org>
Date: Sun, 19 Apr 2020 12:01:46 +0200
Subject: Normalize intersphinx mappings

---
 src/sage/docs/conf.py | 22 +++++++++++-----------
 1 file changed, 11 insertions(+), 11 deletions(-)

From f2087cc70aed56a9ace2c7fccca329ea26ebe8cd Mon Sep 17 00:00:00 2001
From: Antonio Rojas <arojas@archlinux.org>
Date: Sun, 19 Apr 2020 12:07:05 +0200
Subject: Update sphinx to 3.0.1

---
 build/pkgs/sphinx/checksums.ini       | 6 +++---
 build/pkgs/sphinx/package-version.txt | 2 +-
 2 files changed, 4 insertions(+), 4 deletions(-)

From 1fe896979ba545da592a9c8be1179bd82384166b Mon Sep 17 00:00:00 2001
From: Antonio Rojas <arojas@archlinux.org>
Date: Sun, 19 Apr 2020 12:08:15 +0200
Subject: Fix deprecation warnings with sphinx 3

---
 src/sage_setup/docbuild/ext/sage_autodoc.py | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

From 34294f3b6732188d959b38541a3f6b9892d50e6c Mon Sep 17 00:00:00 2001
From: Antonio Rojas <arojas@archlinux.org>
Date: Sun, 19 Apr 2020 12:09:29 +0200
Subject: Port away from BuildEnvironment.frompickle, removed in sphinx 3

---
 src/sage_setup/docbuild/__init__.py | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

From 8cc052d4789a1fc29e51e483169f6f4a15c0da3a Mon Sep 17 00:00:00 2001
From: Antonio Rojas <arojas@archlinux.org>
Date: Sun, 19 Apr 2020 12:13:41 +0200
Subject: Use correct citation domain

---
 src/sage_setup/docbuild/ext/multidocs.py | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

From 7041a6bff42f1f03d832eeed506701c6209cd897 Mon Sep 17 00:00:00 2001
From: Antonio Rojas <arojas@archlinux.org>
Date: Sun, 19 Apr 2020 12:17:12 +0200
Subject: Port to index domain

---
 src/sage_setup/docbuild/ext/multidocs.py | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

From 0705982d22cbdd71e5343dbbdeaa6c3865a9b560 Mon Sep 17 00:00:00 2001
From: Antonio Rojas <arojas@archlinux.org>
Date: Sun, 19 Apr 2020 14:05:51 +0200
Subject: fix WARNING: duplicate object description

---
 src/sage/combinat/permutation.py | 3 ---
 1 file changed, 3 deletions(-)

From 10c7ffd6e1b1f10ec3db39c17c4356741b5fc243 Mon Sep 17 00:00:00 2001
From: Antonio Rojas <arojas@archlinux.org>
Date: Sun, 19 Apr 2020 23:27:59 +0200
Subject: Port to todo domain

---
 src/sage_setup/docbuild/ext/multidocs.py | 13 ++++---------
 1 file changed, 4 insertions(+), 9 deletions(-)

From aa41a308eb3bc215cf0390d2e6fe70d58b83fdcf Mon Sep 17 00:00:00 2001
From: Antonio Rojas <arojas@archlinux.org>
Date: Tue, 21 Apr 2020 12:14:19 +0200
Subject: Fix incomplete port to citation domain

---
 src/sage_setup/docbuild/ext/multidocs.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

From 830a5d5c6dbceaf7505947726a6b36f58b67e6cc Mon Sep 17 00:00:00 2001
From: Antonio Rojas <arojas@archlinux.org>
Date: Tue, 21 Apr 2020 19:00:42 +0200
Subject: Remove non-python code from python code block

---
 src/doc/en/thematic_tutorials/structures_in_coding_theory.rst | 2 --
 1 file changed, 2 deletions(-)

From 0a64b94508bfe0b28750681d463c2aedc25379b3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Thu, 23 Apr 2020 22:27:02 +1200
Subject: Fix the last two doctests

---
 src/sage/misc/sagedoc.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

From 454f2a88102b346c16139382c4af34d0db2ccd8f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Fri, 24 Apr 2020 08:00:45 +1200
Subject: add forgotten doctest fix.

---
 src/sage/docs/conf.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/sage/src/sage_setup/docbuild/ext/multidocs.py
+++ b/sage/src/sage_setup/docbuild/ext/multidocs.py
@@ -52,27 +52,29 @@
         docenv = get_env(app, curdoc)
         if docenv is not None:
             fixpath = lambda path: os.path.join(curdoc, path)
+            todos = docenv.domaindata['todo'].get('todos', dict())
+            citations = docenv.domaindata['citation'].get('citations', dict())
+            indexentries = docenv.domaindata['index'].get('entries', dict())
             logger.info(" %s todos, %s index, %s citations"%(
-                    len(docenv.todo_all_todos),
-                    len(docenv.indexentries),
-                    len(docenv.domaindata["std"]["citations"])
+                    sum(len(t) for t in todos.values()),
+                    len(indexentries),
+                    len(citations)
                     ), nonl=1)
 
             # merge titles
             for t in docenv.titles:
                 env.titles[fixpath(t)] = docenv.titles[t]
             # merge the todo links
-            for dct in docenv.todo_all_todos:
-                dct['docname'] = fixpath(dct['docname'])
-            env.todo_all_todos += docenv.todo_all_todos
+            for dct in todos:
+                env.domaindata['todo']['todos'][fixpath(dct)] = todos[dct]
             # merge the html index links
             newindex = {}
-            for ind in docenv.indexentries:
+            for ind in indexentries:
                 if ind.startswith('sage/'):
-                    newindex[fixpath(ind)] = docenv.indexentries[ind]
+                    newindex[fixpath(ind)] = indexentries[ind]
                 else:
-                    newindex[ind] = docenv.indexentries[ind]
-            env.indexentries.update(newindex)
+                    newindex[ind] = indexentries[ind]
+            env.domaindata['index']['entries'].update(newindex)
             # merge the all_docs links, needed by the js index
             newalldoc = {}
             for ind in docenv.all_docs:
@@ -86,11 +88,10 @@
                 env.metadata[ind] = md
             # merge the citations
             newcite = {}
-            citations = docenv.domaindata["std"]["citations"]
-            for ind, (path, tag, lineno) in six.iteritems(docenv.domaindata["std"]["citations"]):
+            for ind, (path, tag, lineno) in six.iteritems(citations):
                 # TODO: Warn on conflicts
                 newcite[ind] = (fixpath(path), tag, lineno)
-            env.domaindata["std"]["citations"].update(newcite)
+            env.domaindata['citation']['citations'].update(newcite)
             # merge the py:module indexes
             newmodules = {}
             for ind,(modpath,v1,v2,v3) in (
@@ -99,11 +100,11 @@
             env.domaindata['py']['modules'].update(newmodules)
             logger.info(", %s modules"%(len(newmodules)))
     logger.info('... done (%s todos, %s index, %s citations, %s modules)'%(
-            len(env.todo_all_todos),
-            len(env.indexentries),
-            len(env.domaindata["std"]["citations"]),
+            sum(len(t) for t in env.domaindata['todo']['todos'].values()),
+            len(env.domaindata['index']['entries']),
+            len(env.domaindata['citation']['citations']),
             len(env.domaindata['py']['modules'])))
-    write_citations(app, env.domaindata["std"]["citations"])
+    write_citations(app, env.domaindata['citation']['citations'])
 
 
 def get_env(app, curdoc):
@@ -253,7 +254,7 @@
     with open(filename, 'rb') as f:
         cache = cPickle.load(f)
     logger.info("done (%s citations)."%len(cache))
-    cite = env.domaindata["std"]["citations"]
+    cite = env.domaindata['citation'].get('citations', dict())
     for ind, (path, tag, lineno) in six.iteritems(cache):
         if ind not in cite: # don't override local citation
             cite[ind] = (os.path.join("..", path), tag, lineno)
--- a/sage/src/sage/docs/conf.py
+++ b/sage/src/sage/docs/conf.py
@@ -8,6 +8,7 @@
 from docutils.transforms import Transform
 from sphinx.ext.doctest import blankline_re
 from sphinx import highlighting
+import sphinx.ext.intersphinx as intersphinx
 from IPython.lib.lexers import IPythonConsoleLexer, IPyLexer
 
 # If your extensions are in another directory, add it here.
@@ -173,13 +174,8 @@
 
 # Cross-links to other project's online documentation.
 python_version = sys.version_info.major
-intersphinx_mapping = {
-    'python': ('https://docs.python.org/',
-                os.path.join(SAGE_DOC_SRC, "common",
-                             "python{}.inv".format(python_version))),
-    'pplpy': (PPLPY_DOCS, None)}
 
-def set_intersphinx_mappings(app):
+def set_intersphinx_mappings(app, config):
     """
     Add precompiled inventory (the objects.inv)
     """
@@ -190,7 +186,11 @@
         app.config.intersphinx_mapping = {}
         return
 
-    app.config.intersphinx_mapping = intersphinx_mapping
+    app.config.intersphinx_mapping =  {
+    'python': ('https://docs.python.org/',
+                os.path.join(SAGE_DOC_SRC, "common",
+                             "python{}.inv".format(python_version))),
+    'pplpy': (PPLPY_DOCS, None)}
 
     # Add master intersphinx mapping
     dst = os.path.join(invpath, 'objects.inv')
@@ -205,6 +205,7 @@
             dst = os.path.join(invpath, directory, 'objects.inv')
             app.config.intersphinx_mapping[src] = dst
 
+    intersphinx.normalize_intersphinx_mapping(app, config)
 
 # By default document are not master.
 multidocs_is_master = True
@@ -669,11 +670,11 @@
         sage: for line in open(thematic_index).readlines():  # optional - dochtml
         ....:     if "padics" in line:
         ....:         _ = sys.stdout.write(line)
-        <li><a class="reference external" href="../reference/padics/sage/rings/padics/tutorial.html#sage-rings-padics-tutorial" title="(in Sage... Reference Manual: p-Adics v...)"><span>Introduction to the p-adics</span></a></li>
+        <li><p><a class="reference external" href="../reference/padics/sage/rings/padics/tutorial.html#sage-rings-padics-tutorial" title="(in Sage... Reference Manual: p-Adics v...)"><span>Introduction to the p-adics</span></a></p></li>
     """
     debug_inf(app, "???? Trying intersphinx for %s" % node['reftarget'])
     builder = app.builder
-    res =  sphinx.ext.intersphinx.missing_reference(
+    res =  intersphinx.missing_reference(
         app, env, node, contnode)
     if res:
         # Replace absolute links to $SAGE_DOC by relative links: this
@@ -856,11 +857,10 @@
     if app.srcdir.startswith(SAGE_DOC_SRC):
         app.add_config_value('intersphinx_mapping', {}, False)
         app.add_config_value('intersphinx_cache_limit', 5, False)
+        app.connect('config-inited', set_intersphinx_mappings)
+        app.connect('builder-inited', intersphinx.load_mappings)
         # We do *not* fully initialize intersphinx since we call it by hand
         # in find_sage_dangling_links.
         #   app.connect('missing-reference', missing_reference)
         app.connect('missing-reference', find_sage_dangling_links)
-        import sphinx.ext.intersphinx
-        app.connect('builder-inited', set_intersphinx_mappings)
-        app.connect('builder-inited', sphinx.ext.intersphinx.load_mappings)
         app.connect('builder-inited', nitpick_patch_config)
--- a/sage/build/pkgs/sphinx/checksums.ini
+++ b/sage/build/pkgs/sphinx/checksums.ini
@@ -1,4 +1,4 @@
 tarball=Sphinx-VERSION.tar.gz
-sha1=93e715cb5d16d981698caa2d54b3b58596798f5d
-md5=554f7a4e752f48b2601e5ef5ab463346
-cksum=2134440192
+sha1=dd040aedbef765189c6a502c4031df023ae51f46
+md5=111982b6ca04130d4e512f59b2e4402f
+cksum=4119668724
--- a/sage/build/pkgs/sphinx/package-version.txt
+++ b/sage/build/pkgs/sphinx/package-version.txt
@@ -1 +1 @@
-1.8.5.p0
+3.0.1.p0
--- a/sage/src/sage_setup/docbuild/ext/sage_autodoc.py
+++ b/sage/src/sage_setup/docbuild/ext/sage_autodoc.py
@@ -35,14 +35,15 @@
 from docutils.statemachine import ViewList
 
 import sphinx
-from sphinx.ext.autodoc.importer import mock, import_object, get_object_members
+from sphinx.ext.autodoc import mock
+from sphinx.ext.autodoc.importer import import_object, get_object_members, get_module_members
 from sphinx.locale import _, __
 from sphinx.pycode import ModuleAnalyzer
 from sphinx.errors import PycodeError
 from sphinx.util import logging
 from sphinx.util import rpartition, force_decode
 from sphinx.util.docstrings import prepare_docstring
-from sphinx.util.inspect import isdescriptor, safe_getmembers, \
+from sphinx.util.inspect import isdescriptor, \
     safe_getattr, object_description, is_builtin_class_method, \
     isenumattribute, isclassmethod, isstaticmethod, getdoc
 
@@ -536,7 +537,7 @@
 
         # add content from docstrings
         if not no_docstring:
-            encoding = self.analyzer and self.analyzer.encoding
+            encoding = self.analyzer and self.analyzer._encoding
             docstrings = self.get_doc(encoding)
             if not docstrings:
                 # append at least a dummy docstring, so that the event
@@ -882,7 +883,7 @@
             if not hasattr(self.object, '__all__'):
                 # for implicit module members, check __module__ to avoid
                 # documenting imported objects
-                return True, safe_getmembers(self.object)
+                return True, get_module_members(self.object)
             else:
                 memberlist = self.object.__all__
                 # Sometimes __all__ is broken...
@@ -893,7 +894,7 @@
                         '(in module %s) -- ignoring __all__' %
                         (memberlist, self.fullname))
                     # fall back to all members
-                    return True, safe_getmembers(self.object)
+                    return True, get_module_members(self.object)
         else:
             memberlist = self.options.members or []
         ret = []
--- a/sage/src/sage_setup/docbuild/__init__.py
+++ b/sage/src/sage_setup/docbuild/__init__.py
@@ -817,9 +817,13 @@
 
         env_pickle = os.path.join(self._doctrees_dir(), 'environment.pickle')
         try:
-            env = BuildEnvironment.frompickle(env_pickle, FakeApp(self.dir))
-            logger.debug("Opened Sphinx environment: %s", env_pickle)
-            return env
+            with open(env_pickle, 'rb') as f:
+                import pickle
+                env = pickle.load(f)
+                env.app = FakeApp(self.dir)
+                env.config.values = env.app.config.values
+                logger.debug("Opened Sphinx environment: %s", env_pickle)
+                return env
         except IOError as err:
             logger.debug("Failed to open Sphinx environment: %s", err)
 
--- a/sage/src/sage/combinat/permutation.py
+++ b/sage/src/sage/combinat/permutation.py
@@ -415,9 +415,6 @@
         []
         sage: Permutation( [[], []] )
         []
-
-    .. automethod:: Permutation.left_action_product
-    .. automethod:: Permutation.right_action_product
     """
     @staticmethod
     def __classcall_private__(cls, l, check_input = True):
--- a/sage/src/doc/en/thematic_tutorials/structures_in_coding_theory.rst
+++ b/sage/src/doc/en/thematic_tutorials/structures_in_coding_theory.rst
@@ -721,8 +721,6 @@
 
 .. CODE-BLOCK:: python
 
-    :class:`sage.coding.repetition_code.BinaryRepetitionCode <sage.coding.repetition_code.BinaryRepetitionCode>`
-    #the line above creates a link to the class in the html documentation of coding theory library
     from sage.coding.repetition_code import BinaryRepetitionCode
 
 ``encoders_catalog.py`` (continued):
--- a/sage/src/sage/misc/sagedoc.py
+++ b/sage/src/sage/misc/sagedoc.py
@@ -24,7 +24,7 @@
     ....:     for line in fobj:
     ....:         if "#sage.symbolic.expression.Expression.numerical_approx" in line:
     ....:             print(line)
-    <code class="descname">numerical_approx</code><span class="sig-paren">(</span><em>prec=None</em>, <em>digits=None</em>, <em>algorithm=None</em><span class="sig-paren">)</span>...
+    <code class="sig-name descname">numerical_approx</code><span class="sig-paren">(</span><em class="sig-param">...
 
 Check that sphinx is not imported at Sage start-up::
 
